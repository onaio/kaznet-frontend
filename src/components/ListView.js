// This component renders a list using Bootstrap 4 tables
import _ from "lodash";
import React, { Component } from "react";
import {
  Table,
  Pagination,
  PaginationItem,
  Dropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
  NavLink,
  Button,
  Modal,
  ModalBody,
  ModalHeader,
  Input,
  Col,
  Row,
  FormGroup,
  Form
} from "reactstrap";
import { Link } from "react-router-dom";
import * as constants from "../constants";

import "./ListView.css";

export default class ListView extends Component {
  render() {
    if (this.props.sortField) {
      const sortField = this.props.sortField;
      var sortOrder = constants.SORT_ASC;
      if (this.props.sortOrder) {
        sortOrder = this.props.sortOrder;
      }
      const rowsArray = _.values(this.props.rowsById);
      this.sortedRowsArray = _.orderBy(
        rowsArray,
        function(e) {
          return e.attributes[sortField];
        },
        [sortOrder]
      );
    }

    const statuses = ["", ...constants.TASK_STATUSES];
    const statusArr = statuses.map(s => {
      let status = "";
      switch (s) {
        case constants.TASK_ACTIVE:
          status = constants.ACTIVE;
          break;
        case constants.TASK_DEACTIVATED:
          status = constants.DEACTIVATED;
          break;
        case constants.TASK_EXPIRED:
          status = constants.EXPIRED;
          break;
        case constants.TASK_DRAFT:
          status = constants.DRAFT;
          break;
        case constants.TASK_ARCHIVED:
          status = constants.ARCHIVED;
          break;
        case constants.TASK_SCHEDULED:
          status = constants.SCHEDULED;
          break;
        case constants.TASK_ALL:
          status = constants.ALL;
          break;
        default:
          status = "";
      }
      return (
        <DropdownItem
          className={`${s === this.props.taskStatus ? "active-task" : ""}`}
          key={s}
        >
          <NavLink className="nav-link" key={s} data-key={s}>
            {status}
          </NavLink>
        </DropdownItem>
      );
    });
    return (
      <div>
        {this.props.downloadModalHandler && (
          <div>
            <Modal
              isOpen={this.props.modalState}
              toggle={this.props.downloadModalHandler}
              className={this.props.className}
            >
              <ModalHeader toggle={this.props.downloadModalHandler}>
                Submissions of {this.props.userName}
              </ModalHeader>
              <ModalBody>
                <p className="font-weight-normal">Export Submissions </p>
                <Form onSubmit={this.props.onFormSubmit}>
                  <FormGroup className="row">
                    <Row>
                      <p className="text-center align-middle ml-4">From:</p>
                      <Col md={{ size: 5 }}>
                        <Input
                          name="start"
                          type="date"
                          bsSize="md"
                          placeholder="Start Date"
                          aria-label="start"
                          className={`time-picker`}
                          onChange={this.props.handleDateChanges}
                        />
                      </Col>
                      <p className="text-center align-middle">to</p>
                      <Col md="5">
                        <Input
                          name="end"
                          type="date"
                          bsSize="md"
                          placeholder="End Date"
                          aria-label="end"
                          className={`time-picker`}
                          onChange={this.props.handleDateChanges}
                        />
                      </Col>
                    </Row>
                  </FormGroup>
                  <FormGroup className="row">
                    <Col md={{ size: 5, offset: 1 }}>
                      <Button
                        className="btn btn-secondary btn-block"
                        color="secondary"
                        onClick={this.props.downloadModalHandler}
                      >
                        Cancel
                      </Button>
                    </Col>
                    <Col md={{ size: 5 }}>
                      <Button
                        type="submit"
                        className="btn btn-primary btn-block"
                        color="secondary"
                        onClick={this.props.downloadModalHandler}
                      >
                        Export
                      </Button>
                    </Col>
                  </FormGroup>
                </Form>
              </ModalBody>
            </Modal>
          </div>
        )}

        <Table bordered className="kaznet-table">
          <thead>
            <tr>
              {this.props.isTaskPage ? (
                <th>
                  <Dropdown
                    isOpen={this.props.isOpen}
                    toggle={e => this.props.handleChange(e)}
                  >
                    <DropdownToggle caret tag="span" className="nav-link">
                      Status
                    </DropdownToggle>
                    <DropdownMenu>{statusArr}</DropdownMenu>
                  </Dropdown>
                </th>
              ) : null}
              {this.props.renderHeaders()}
            </tr>
          </thead>
          <tbody>
            {this.props.sortField
              ? _.map(this.sortedRowsArray, this.renderRowObject.bind(this))
              : _.map(this.props.rowsIdArray, this.renderRowById.bind(this))}
          </tbody>
        </Table>
        {this.props.totalPages > 1 ? this.renderPagination() : null}
      </div>
    );
  }

  renderRowById(rowId) {
    return (
      <tr key={rowId}>
        {this.props.renderRow(
          _.get(this.props.rowsById, rowId),
          this.props.downloadModalHandler,
          this.props.setUserDetails
        )}
      </tr>
    );
  }

  renderRowObject(rowObject) {
    return <tr key={rowObject.id}>{this.props.renderRow(rowObject)}</tr>;
  }

  renderPagination() {
    return (
      <Pagination aria-label="Page navigation example">
        <PaginationItem disabled={this.props.pageLinks.first ? false : true}>
          <Link
            to={
              this.props.pageLinks.first
                ? `/${this.props.endpoint}/?search=${
                    !this.props.searchVal || this.props.searchVal === undefined
                      ? ""
                      : this.props.searchVal
                  }&status=${
                    !this.props.taskStatus ||
                    this.props.taskStatus === undefined
                      ? ""
                      : this.props.taskStatus
                  }&page=${this.props.firstPage}`
                : "#"
            }
            className="page-link"
            aria-label="First"
          >
            First
          </Link>
        </PaginationItem>
        <PaginationItem disabled={this.props.pageLinks.prev ? false : true}>
          <Link
            to={
              this.props.pageLinks.prev
                ? `/${this.props.endpoint}/?search=${
                    !this.props.searchVal ? "" : this.props.searchVal
                  }&status=${
                    !this.props.taskStatus ? "" : this.props.taskStatus
                  }&page=${this.props.currentPage - 1}`
                : "#"
            }
            className="page-link"
            aria-label="Previous"
          >
            &laquo;
          </Link>
        </PaginationItem>
        <PaginationItem disabled={this.props.pageLinks.next ? false : true}>
          <Link
            to={
              this.props.pageLinks.next
                ? `/${this.props.endpoint}/?search=${
                    !this.props.searchVal ? "" : this.props.searchVal
                  }&status=${
                    !this.props.taskStatus ? "" : this.props.taskStatus
                  }&page=${this.props.currentPage + 1}`
                : "#"
            }
            className="page-link"
            aria-label="Next"
          >
            &raquo;
          </Link>
        </PaginationItem>
        <PaginationItem disabled={this.props.pageLinks.last ? false : true}>
          <Link
            to={
              this.props.pageLinks.last
                ? `/${this.props.endpoint}/?search=${
                    !this.props.searchVal ? "" : this.props.searchVal
                  }&status=${
                    !this.props.taskStatus ? "" : this.props.taskStatus
                  }&page=${
                    !this.props.lastPage || this.props.lastPage === undefined
                      ? 1
                      : this.props.lastPage
                  }`
                : "#"
            }
            className="page-link"
            aria-label="Last"
          >
            Last
          </Link>
        </PaginationItem>
      </Pagination>
    );
  }
}
