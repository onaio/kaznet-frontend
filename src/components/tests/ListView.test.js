// Test ListView
import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";

import ErrorBoundary from "../ErrorBoundary";
import ElementMap from "../../containers/ElementMap";
import ListView from "../ListView";
const history = createBrowserHistory();

describe("components/ListView", () => {
  it("renders without crashing", () => {
    const rowsById = { "0": "Bob", "1": "Jane" };
    const rowsIdArray = ["0", "1"];

    const currentPage = 1;
    const totalPages = 2;
    const pageLinks = {
      first: null,
      last: null,
      prev: null,
      next: null
    };

    const renderHeader = function() {
      const headerItems = ["Name"];
      return <ElementMap items={headerItems} HTMLTag="th" />;
    };

    const renderRow = function(row) {
      const rowItems = [row];
      return <ElementMap items={rowItems} HTMLTag="td" />;
    };

    shallow(
      <ListView
        renderHeaders={renderHeader}
        rowsIdArray={rowsIdArray}
        rowsById={rowsById}
        pageLinks={pageLinks}
        currentPage={currentPage}
        totalPages={totalPages}
        renderRow={renderRow}
      />
    );
  });

  it("renders a table", () => {
    const rowsById = { "0": "Bob", "1": "Jane" };
    const rowsIdArray = ["0", "1"];
    const currentPage = 1;
    const totalPages = 2;
    const pageLinks = {
      first: null,
      last: null,
      prev: null,
      next: null
    };

    const renderHeader = function() {
      const headerItems = ["Name"];
      return <ElementMap items={headerItems} HTMLTag="th" />;
    };

    const renderRow = function(row) {
      const rowItems = [row];
      return <ElementMap items={rowItems} HTMLTag="td" />;
    };

    const wrapper = mount(
      <Router history={history}>
        <ListView
          renderHeaders={renderHeader}
          rowsIdArray={rowsIdArray}
          rowsById={rowsById}
          renderRow={renderRow}
          pageLinks={pageLinks}
          totalPages={totalPages}
          currentPage={currentPage}
        />
      </Router>
    );

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.unmount();
  });
});
