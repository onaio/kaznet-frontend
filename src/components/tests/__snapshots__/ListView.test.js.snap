// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ListView does not render pagination when totalPages <= 1 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ListView
    currentPage={1}
    endpoint="http://10.20.25.28:8000/api/v1"
    firstPage={1}
    lastpage={1}
    pageLinks={
      Object {
        "first": "http://localhost:8000/api/v1/clients/?page=1",
        "last": null,
        "next": null,
        "prev": null,
      }
    }
    renderHeaders={[Function]}
    renderRow={[Function]}
    rowsById={
      Object {
        "0": "Bob",
        "1": "Jane",
      }
    }
    rowsIdArray={
      Array [
        "0",
        "1",
      ]
    }
    totalPages={1}
  >
    <div>
      <Table
        bordered={true}
        className="kaznet-table"
        responsiveTag="div"
        tag="table"
      >
        <table
          className="kaznet-table table table-bordered"
        >
          <thead>
            <tr>
              <ElementMap
                HTMLTag="th"
                items={
                  Array [
                    "Name",
                  ]
                }
              >
                <th
                  key="0"
                >
                  Name
                </th>
              </ElementMap>
            </tr>
          </thead>
          <tbody>
            <tr
              key="0"
            >
              <ElementMap
                HTMLTag="td"
                items={
                  Array [
                    "Bob",
                  ]
                }
              >
                <td
                  key="0"
                >
                  Bob
                </td>
              </ElementMap>
            </tr>
            <tr
              key="1"
            >
              <ElementMap
                HTMLTag="td"
                items={
                  Array [
                    "Jane",
                  ]
                }
              >
                <td
                  key="0"
                >
                  Jane
                </td>
              </ElementMap>
            </tr>
          </tbody>
        </table>
      </Table>
    </div>
  </ListView>
</Router>
`;

exports[`components/ListView renders a table 1`] = `
<ErrorBoundary>
  <ListView
    renderHeaders={[Function]}
    renderRow={[Function]}
    rowsById={
      Object {
        "0": "Bob",
        "1": "Jane",
      }
    }
    rowsIdArray={
      Array [
        "0",
        "1",
      ]
    }
  >
    <div>
      <Table
        bordered={true}
        className="kaznet-table"
        responsiveTag="div"
        tag="table"
      >
        <table
          className="kaznet-table table table-bordered"
        >
          <thead>
            <tr>
              <ElementMap
                HTMLTag="th"
                items={
                  Array [
                    "Name",
                  ]
                }
              >
                <th
                  key="0"
                >
                  Name
                </th>
              </ElementMap>
            </tr>
          </thead>
          <tbody>
            <tr
              key="0"
            >
              <ElementMap
                HTMLTag="td"
                items={
                  Array [
                    "Bob",
                  ]
                }
              >
                <td
                  key="0"
                >
                  Bob
                </td>
              </ElementMap>
            </tr>
            <tr
              key="1"
            >
              <ElementMap
                HTMLTag="td"
                items={
                  Array [
                    "Jane",
                  ]
                }
              >
                <td
                  key="0"
                >
                  Jane
                </td>
              </ElementMap>
            </tr>
          </tbody>
        </table>
      </Table>
    </div>
  </ListView>
</ErrorBoundary>
`;

exports[`components/ListView renders pagination when totalPages > 1 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ListView
    currentPage={1}
    endpoint="http://10.20.25.28:8000/api/v1"
    firstPage={1}
    lastpage={2}
    pageLinks={
      Object {
        "first": "http://localhost:8000/api/v1/clients/?page=1",
        "last": "http://localhost:8000/api/v1/clients/?page=2",
        "next": "http://localhost:8000/api/v1/clients/?page=2",
        "prev": null,
      }
    }
    renderHeaders={[Function]}
    renderRow={[Function]}
    rowsById={
      Object {
        "0": "Bob",
        "1": "Jane",
      }
    }
    rowsIdArray={
      Array [
        "0",
        "1",
      ]
    }
    totalPages={2}
  >
    <div>
      <Table
        bordered={true}
        className="kaznet-table"
        responsiveTag="div"
        tag="table"
      >
        <table
          className="kaznet-table table table-bordered"
        >
          <thead>
            <tr>
              <ElementMap
                HTMLTag="th"
                items={
                  Array [
                    "Name",
                  ]
                }
              >
                <th
                  key="0"
                >
                  Name
                </th>
              </ElementMap>
            </tr>
          </thead>
          <tbody>
            <tr
              key="0"
            >
              <ElementMap
                HTMLTag="td"
                items={
                  Array [
                    "Bob",
                  ]
                }
              >
                <td
                  key="0"
                >
                  Bob
                </td>
              </ElementMap>
            </tr>
            <tr
              key="1"
            >
              <ElementMap
                HTMLTag="td"
                items={
                  Array [
                    "Jane",
                  ]
                }
              >
                <td
                  key="0"
                >
                  Jane
                </td>
              </ElementMap>
            </tr>
          </tbody>
        </table>
      </Table>
      <Pagination
        aria-label="Page navigation example"
        listTag="ul"
        tag="nav"
      >
        <nav
          aria-label="Page navigation example"
          className=""
        >
          <ul
            className="pagination"
          >
            <PaginationItem
              disabled={false}
              tag="li"
            >
              <li
                className="page-item"
              >
                <Link
                  aria-label="First"
                  className="page-link"
                  replace={false}
                  to="/http://10.20.25.28:8000/api/v1/?search=&page=1"
                >
                  <a
                    aria-label="First"
                    className="page-link"
                    href="/http://10.20.25.28:8000/api/v1/?search=&page=1"
                    onClick={[Function]}
                  >
                    First
                  </a>
                </Link>
              </li>
            </PaginationItem>
            <PaginationItem
              disabled={true}
              tag="li"
            >
              <li
                className="page-item disabled"
              >
                <Link
                  aria-label="Previous"
                  className="page-link"
                  replace={false}
                  to="#"
                >
                  <a
                    aria-label="Previous"
                    className="page-link"
                    href="/"
                    onClick={[Function]}
                  >
                    «
                  </a>
                </Link>
              </li>
            </PaginationItem>
            <PaginationItem
              disabled={false}
              tag="li"
            >
              <li
                className="page-item"
              >
                <Link
                  aria-label="Next"
                  className="page-link"
                  replace={false}
                  to="/http://10.20.25.28:8000/api/v1/?search=&page=2"
                >
                  <a
                    aria-label="Next"
                    className="page-link"
                    href="/http://10.20.25.28:8000/api/v1/?search=&page=2"
                    onClick={[Function]}
                  >
                    »
                  </a>
                </Link>
              </li>
            </PaginationItem>
            <PaginationItem
              disabled={false}
              tag="li"
            >
              <li
                className="page-item"
              >
                <Link
                  aria-label="Last"
                  className="page-link"
                  replace={false}
                  to="/http://10.20.25.28:8000/api/v1/?search=&page=1"
                >
                  <a
                    aria-label="Last"
                    className="page-link"
                    href="/http://10.20.25.28:8000/api/v1/?search=&page=1"
                    onClick={[Function]}
                  >
                    Last
                  </a>
                </Link>
              </li>
            </PaginationItem>
          </ul>
        </nav>
      </Pagination>
    </div>
  </ListView>
</Router>
`;
